pynicotine/transfers.py

```diff
@@ -1036,12 +1036,12 @@ def file_download_init(self, msg):
                     md5sum = md5()
                     md5sum.update((i.filename + i.user).encode('utf-8'))
 
-                    base_name, extension = os.path.splitext(clean_file(i.filename.replace('/', '\\').split('\\')[-1]))
```

Comment:
I have a feeling that `clean_file()` is required because it is possible the filename may include characters which the OS does not allow.

Reply:
Sure, I can add that. I was wondering, does Nicotine+ replace any of the characters in [ILLEGALPATHCHARS](https://github.com/nicotine-plus/nicotine-plus/blob/cbd09acb6506f6e58c192702e9a41a244d790879/pynicotine/utils.py#L41) on moving the complete file to the download directory? What about in displaying filenames in the GUI? Because I use a system that supports any byte except '/' and NUL, I would want to avoid replacing characters whenever possible.

Reply:
> Sure, I can add that

Thanks, better to be on the safe side, especially considering this is only a temporary file nobody cares if the filename is not precisely correct.

> does Nicotine+ replace any of the characters in [ILLEGALPATHCHARS](https://github.com/nicotine-plus/nicotine-plus/blob/cbd09acb6506f6e58c192702e9a41a244d790879/pynicotine/utils.py#L41) on moving the complete file to the download directory?

Yes, `clean_file()` uses [ILLEGALFILECHARS](https://github.com/nicotine-plus/nicotine-plus/blob/cbd09acb6506f6e58c192702e9a41a244d790879/pynicotine/utils.py#L42) which does contain all of the characters in ILLEGALPATHCHARS and '\\', '/' which is called from `get_download_destination()`:

https://github.com/nicotine-plus/nicotine-plus/blob/cbd09acb6506f6e58c192702e9a41a244d790879/pynicotine/transfers.py#L1755-L1761

> I would want to avoid replacing characters whenever possible

Perhaps the constants containing the characters could be customized, maybe with `sys.platform` or some other logic in utils.py to determine the exact capabilities of the filesystem. However, this would give different results on different systems leading to file duplication or other confusions, whereas at the moment it is the same limitations for everybody.

Reply:
> Perhaps the constants containing the characters could be customized, maybe with `sys.platform` or some other logic in utils.py to determine the exact capabilities of the filesystem. However, this would give different results on different systems leading to file duplication or other confusions, whereas at the moment it is the same limitations for everybody.
> 
> Another way to do it might be to always attempt to first create the file with the raw name, and then catch an exception to fall back to the clean_file name if required.

I'll implement it privately or maybe in a public fork then, and later submit upstream in case it's a welcome change for everybody to use.